File: src/lib/blackoutService.ts
import { WithId } from 'mongodb';
import { addMinutes, differenceInMinutes, isAfter, isBefore, isValid, startOfDay } from 'date-fns';
import { getDb } from './db';
import { BlackoutSchedule, BlackoutInterval } from '@/types';

const COLLECTION_NAME = 'apagones';

type BlackoutIntervalDocument = {
  start: Date;
  end: Date;
  durationMinutes: number;
};

type BlackoutDocument = {
  date: Date;
  intervals: BlackoutIntervalDocument[];
  province?: string;
  municipality?: string;
  notes?: string;
  createdAt: Date;
  updatedAt: Date;
};

export interface SaveBlackoutPayload {
  date?: string;
  intervals?: Array<{
    start?: string;
    end?: string;
  }>;
  province?: string;
  municipality?: string;
  notes?: string;
}

function mapInterval(doc: BlackoutIntervalDocument): BlackoutInterval {
  return {
    start: doc.start.toISOString(),
    end: doc.end.toISOString(),
    durationMinutes: doc.durationMinutes,
  };
}

function mapSchedule(doc: WithId<BlackoutDocument>): BlackoutSchedule {
  return {
    _id: doc._id.toHexString(),
    date: doc.date.toISOString(),
    intervals: doc.intervals.map(mapInterval),
    province: doc.province,
    municipality: doc.municipality,
    notes: doc.notes,
    createdAt: doc.createdAt.toISOString(),
    updatedAt: doc.updatedAt.toISOString(),
  };
}

function ensureValidDate(dateStr?: string): Date {
  if (!dateStr) {
    throw new Error('La fecha del día es obligatoria.');
  }
  const date = new Date(dateStr);
  if (!isValid(date)) {
    throw new Error('La fecha proporcionada es inválida.');
  }
  return startOfDay(date);
}

function ensureValidIntervals(date: Date, intervals?: SaveBlackoutPayload['intervals']): BlackoutIntervalDocument[] {
  if (!intervals || intervals.length === 0) {
    throw new Error('Se requiere al menos un intervalo de apagón.');
  }

  const normalized: BlackoutIntervalDocument[] = intervals.map((interval, index) => {
    if (!interval.start || !interval.end) {
      throw new Error(`El intervalo #${index + 1} debe incluir hora de inicio y fin.`);
    }
    const start = new Date(interval.start);
    const end = new Date(interval.end);

    if (!isValid(start) || !isValid(end)) {
      throw new Error(`El intervalo #${index + 1} tiene fechas inválidas.`);
    }

    if (isAfter(start, end) || start.getTime() === end.getTime()) {
      throw new Error(`El intervalo #${index + 1} debe tener fin posterior al inicio.`);
    }

    if (!isSameDayRange(date, start, end)) {
      throw new Error(`El intervalo #${index + 1} debe pertenecer al mismo día indicado.`);
    }

    const durationMinutes = differenceInMinutes(end, start);
    if (durationMinutes < 15) {
      throw new Error(`El intervalo #${index + 1} debe durar al menos 15 minutos.`);
    }

    return {
      start,
      end,
      durationMinutes,
    };
  });

  normalized.sort((a, b) => a.start.getTime() - b.start.getTime());

  for (let i = 1; i < normalized.length; i++) {
    const prev = normalized[i - 1];
    const current = normalized[i];
    if (!isBefore(prev.end, current.start)) {
      throw new Error('Los intervalos de apagón no pueden solaparse.');
    }
  }

  return normalized;
}

function isSameDayRange(reference: Date, start: Date, end: Date): boolean {
  const rangeStart = startOfDay(reference);
  const rangeEnd = addMinutes(rangeStart, 24 * 60);
  return (
    !isBefore(start, rangeStart) &&
    isBefore(start, rangeEnd) &&
    !isAfter(end, rangeEnd)
  );
}

export async function saveBlackoutSchedule(payload: SaveBlackoutPayload): Promise<BlackoutSchedule> {
  const date = ensureValidDate(payload.date);
  const intervals = ensureValidIntervals(date, payload.intervals);

  const db = await getDb();
  const collection = db.collection<BlackoutDocument>(COLLECTION_NAME);

  const now = new Date();

  const result = await collection.findOneAndUpdate(
    { date },
    {
      $set: {
        date,
        intervals,
        province: payload.province?.trim() || undefined,
        municipality: payload.municipality?.trim() || undefined,
        notes: payload.notes?.trim() || undefined,
        updatedAt: now,
      },
      $setOnInsert: {
        createdAt: now,
      },
    },
    {
      upsert: true,
      returnDocument: 'after',
    }
  );

  if (!result) {
    throw new Error('No se pudo guardar el horario de apagón.');
  }

  return mapSchedule(result);
}

export interface ListBlackoutsOptions {
  from?: string;
  to?: string;
  limit?: number;
}

export async function getBlackoutByDate(dateStr: string): Promise<BlackoutSchedule | null> {
  const date = ensureValidDate(dateStr);
  const db = await getDb();
  const collection = db.collection<BlackoutDocument>(COLLECTION_NAME);
  const doc = await collection.findOne({ date });
  return doc ? mapSchedule(doc) : null;
}

export async function listBlackouts(options: ListBlackoutsOptions = {}): Promise<BlackoutSchedule[]> {
  const db = await getDb();
  const collection = db.collection<BlackoutDocument>(COLLECTION_NAME);

  const query: Record<string, unknown> = {};
  const filters: Record<string, unknown> = {};

  if (options.from) {
    const fromDate = startOfDay(new Date(options.from));
    if (!isValid(fromDate)) {
      throw new Error('El parámetro "from" es inválido.');
    }
    filters.$gte = fromDate;
  }

  if (options.to) {
    const toDate = startOfDay(new Date(options.to));
    if (!isValid(toDate)) {
      throw new Error('El parámetro "to" es inválido.');
    }
    filters.$lte = toDate;
  }

  if (Object.keys(filters).length > 0) {
    query.date = filters;
  }

  const cursor = collection
    .find(query)
    .sort({ date: 1 });

  if (options.limit && options.limit > 0) {
    cursor.limit(options.limit);
  }

  const docs = await cursor.toArray();
  return docs.map(mapSchedule);
}

export async function getBlackoutsForRange(from: Date, to: Date): Promise<BlackoutSchedule[]> {
  const db = await getDb();
  const collection = db.collection<BlackoutDocument>(COLLECTION_NAME);

  const docs = await collection
    .find({
      date: {
        $gte: startOfDay(from),
        $lte: startOfDay(to),
      },
    })
    .sort({ date: 1 })
    .toArray();

  return docs.map(mapSchedule);
}

export async function deleteBlackout(dateStr: string): Promise<boolean> {
  const date = ensureValidDate(dateStr);
  const db = await getDb();
  const collection = db.collection<BlackoutDocument>(COLLECTION_NAME);
  const { deletedCount } = await collection.deleteOne({ date });
  return deletedCount === 1;
}

File: backend/app/services/blackout_service.py
"""
Blackout schedule service mirroring the original Next.js logic.
"""
from __future__ import annotations

from datetime import datetime, timedelta
from typing import Any, Dict, List, Optional

from bson import ObjectId
from pymongo.collection import Collection

from app.database import get_database

COLLECTION_NAME = "apagones"


def _collection() -> Collection:
    return get_database()[COLLECTION_NAME]


def _start_of_day(date: datetime) -> datetime:
    return date.replace(hour=0, minute=0, second=0, microsecond=0)


def _end_of_day(date: datetime) -> datetime:
    return _start_of_day(date) + timedelta(days=1)


def _parse_iso(value: str) -> datetime:
    if value.endswith("Z"):
        value = value[:-1] + "+00:00"
    return datetime.fromisoformat(value)


def _ensure_valid_date(date_str: Optional[str]) -> datetime:
    if not date_str:
        raise ValueError("La fecha del día es obligatoria.")
    try:
        date = _parse_iso(date_str) if "T" in date_str else datetime.fromisoformat(date_str)
    except ValueError as exc:  # pragma: no cover - invalid iso
        raise ValueError("La fecha proporcionada es inválida.") from exc
    return _start_of_day(date)


def _map_interval(doc: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "start": doc["start"].isoformat(),
        "end": doc["end"].isoformat(),
        "durationMinutes": doc.get("durationMinutes"),
    }


def _map_blackout(doc: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "_id": str(doc["_id"]),
        "date": doc["date"].isoformat(),
        "intervals": [_map_interval(interval) for interval in doc.get("intervals", [])],
        "province": doc.get("province"),
        "municipality": doc.get("municipality"),
        "notes": doc.get("notes"),
        "createdAt": doc.get("createdAt").isoformat() if doc.get("createdAt") else None,
        "updatedAt": doc.get("updatedAt").isoformat() if doc.get("updatedAt") else None,
    }


def _ensure_valid_intervals(date: datetime, intervals: Optional[List[Dict[str, Any]]]) -> List[Dict[str, Any]]:
    if not intervals:
        raise ValueError("Se requiere al menos un intervalo de apagón.")

    normalized: List[Dict[str, Any]] = []
    for idx, interval in enumerate(intervals):
        start_str = interval.get("start")
        end_str = interval.get("end")
        if not start_str or not end_str:
            raise ValueError(f"El intervalo #{idx + 1} debe incluir hora de inicio y fin.")
        try:
            start = _parse_iso(start_str)
            end = _parse_iso(end_str)
        except ValueError as exc:
            raise ValueError(f"El intervalo #{idx + 1} tiene fechas inválidas.") from exc

        if not start < end:
            raise ValueError(f"El intervalo #{idx + 1} debe tener fin posterior al inicio.")

        range_start = _start_of_day(date)
        range_end = _end_of_day(date)
        if not (range_start <= start < range_end and range_start < end <= range_end):
            raise ValueError(f"El intervalo #{idx + 1} debe pertenecer al mismo día indicado.")

        duration_minutes = int((end - start).total_seconds() / 60)
        if duration_minutes < 15:
            raise ValueError(f"El intervalo #{idx + 1} debe durar al menos 15 minutos.")

        normalized.append(
            {
                "start": start,
                "end": end,
                "durationMinutes": duration_minutes,
            }
        )

    normalized.sort(key=lambda item: item["start"])
    for prev, current in zip(normalized, normalized[1:]):
        if not prev["end"] < current["start"]:
            raise ValueError("Los intervalos de apagón no pueden solaparse.")

    return normalized


def _object_id(blackout_id: str) -> ObjectId:
    try:
        return ObjectId(blackout_id)
    except Exception as exc:  # pragma: no cover
        raise ValueError("Identificador de apagón inválido.") from exc


def save_blackout_schedule(payload: Dict[str, Any]) -> Dict[str, Any]:
    date = _ensure_valid_date(payload.get("date"))
    intervals = _ensure_valid_intervals(date, payload.get("intervals"))
    now = datetime.utcnow()

    result = _collection().find_one_and_update(
        {"date": date},
        {
            "$set": {
                "date": date,
                "intervals": intervals,
                "province": payload.get("province") or None,
                "municipality": payload.get("municipality") or None,
                "notes": payload.get("notes") or None,
                "updatedAt": now,
            },
            "$setOnInsert": {"createdAt": now},
        },
        upsert=True,
        return_document=True,
    )

    if not result:
        raise RuntimeError("No se pudo guardar el horario de apagón.")
    return _map_blackout(result)


def update_blackout_schedule(blackout_id: str, payload: Dict[str, Any]) -> Dict[str, Any]:
    date = _ensure_valid_date(payload.get("date"))
    intervals = _ensure_valid_intervals(date, payload.get("intervals"))
    now = datetime.utcnow()

    result = _collection().find_one_and_update(
        {"_id": _object_id(blackout_id)},
        {
            "$set": {
                "date": date,
                "intervals": intervals,
                "province": payload.get("province") or None,
                "municipality": payload.get("municipality") or None,
                "notes": payload.get("notes") or None,
                "updatedAt": now,
            }
        },
        return_document=True,
    )

    if not result:
        raise ValueError("No se encontró el horario de apagón solicitado.")
    return _map_blackout(result)


def list_blackouts(from_date: Optional[str] = None, to_date: Optional[str] = None, limit: Optional[int] = None) -> List[Dict[str, Any]]:
    filters: Dict[str, Any] = {}
    if from_date or to_date:
        filters["date"] = {}
        if from_date:
            filters["date"]["$gte"] = _start_of_day(_ensure_valid_date(from_date))
        if to_date:
            filters["date"]["$lte"] = _start_of_day(_ensure_valid_date(to_date))
        if not filters["date"]:
            filters.pop("date")

    cursor = _collection().find(filters).sort("date", 1)
    if limit and limit > 0:
        cursor = cursor.limit(limit)
    return [_map_blackout(doc) for doc in cursor]


def get_blackout(blackout_id: str) -> Optional[Dict[str, Any]]:
    doc = _collection().find_one({"_id": _object_id(blackout_id)})
    return _map_blackout(doc) if doc else None


def get_blackout_by_date(date_str: str) -> Optional[Dict[str, Any]]:
    date = _ensure_valid_date(date_str)
    doc = _collection().find_one({"date": date})
    return _map_blackout(doc) if doc else None


def get_blackouts_for_range(start: datetime, end: datetime) -> List[Dict[str, Any]]:
    docs = (
        _collection()
        .find({"date": {"$gte": _start_of_day(start), "$lte": _start_of_day(end)}})
        .sort("date", 1)
    )
    return [_map_blackout(doc) for doc in docs]


def delete_blackout(blackout_id: str) -> bool:
    result = _collection().delete_one({"_id": _object_id(blackout_id)})
    return result.deleted_count == 1

